### PR TITLE
cmd/update-report: improve formula file validation

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Currently, ruby files that are not in the Cask directory are considered to be formulae if a Formula or HomebrewFormula directory doesn't exist  which doesn't make sense. We know that these should only be in a few directories so we can check for that explicitly.

Beyond that the `Tap#cask_file?` and `Tap.formula_file?` methods were only used inside update-report so it doesn't make sense to turn them into pathnames and expand things when we know that each string will be a relative path from a tap that we can just check with a regex.

This change will stop other tap changes like new commands or changes to other directories like lib/ from showing up as new formulae.

I tried opening a PR for this a long time ago but I got busy with other things and it got closed by the stale bot.

- https://github.com/Homebrew/brew/pull/15489

### CODE REVIEW

**Comment by apainintheneck** [comment]
Here's an example of the unexpected behavior.

```console
/u/l/Homebrew (master|✔) $ brew update
==> Updating Homebrew...
Updated 1 tap (apainintheneck/dev-utils).
==> New Formulae
apainintheneck/dev-utils/startup-stats
==> Outdated Formulae
gh                                   luajit                               python-setuptools                    shellcheck

You have 4 outdated formulae installed.
You can upgrade them with brew upgrade
or list them with brew outdated.
```

In this case, `startup-stats` is a new command not a new formula.

File: Library/Homebrew/test/tap_spec.rb [hunk 1]
```diff
@@ -201,7 +201,6 @@ def setup_completion(link:)
     expect(homebrew_foo_tap.tap_migrations).to eq("removed-formula" => "homebrew/foo")
     expect(homebrew_foo_tap.command_files).to eq([cmd_file])
     expect(homebrew_foo_tap.to_hash).to be_a(Hash)
-    expect(homebrew_foo_tap).to have_formula_file(formula_file)
```
Comment (apainintheneck) [inline Library/Homebrew/test/tap_spec.rb:204]: I don't remember why I removed this in the old PR...

File: Library/Homebrew/tap.rb [hunk 2]
```diff
@@ -753,34 +753,32 @@ def cask_files_by_name
...
   end
-  private :ruby_file?
+  private :formula_file_regex
 
-  # Check whether the given path would present a {Formula} file in this {Tap}.
-  # Accepts either an absolute path or a path relative to this {Tap}'s path.
-  sig { params(file: T.any(String, Pathname)).returns(T::Boolean) }
+  # accepts the relative path of a file from {Tap}'s path
+  sig { params(file: String).returns(T::Boolean) }
   def formula_file?(file)
```
Comment (apainintheneck) [inline Library/Homebrew/tap.rb:771]: This method also gets used by `brew-test-bot` so I'll need to make sure that is passing a String here before merging.

File: Library/Homebrew/tap.rb [hunk 3]
```diff
@@ -753,34 +753,32 @@ def cask_files_by_name
...
-    file.extname == ".rb"
+  sig { returns(Regexp) }
+  def formula_file_regex
+    @formula_file_regex ||= case formula_dir.basename.to_s
+    when "Formula"
+      %r{^Formula(/[^/]+)+\.rb$}
+    when "HomebrewFormula"
+      %r{^HomebrewFormula(/[^/]+)+\.rb$}
+    else
+      %r{^[^/]+\.rb$}
```
Comment (apainintheneck) [inline Library/Homebrew/tap.rb:764]: I should just switch on all of the paths from the `potential_formula_dirs` method instead of using the basename here.

File: Library/Homebrew/tap.rb [hunk 4]
```diff
@@ -753,34 +753,32 @@ def cask_files_by_name
...
 
-  # Check whether the file has a Ruby extension.
-  sig { params(file: Pathname).returns(T::Boolean) }
-  def ruby_file?(file)
-    file.extname == ".rb"
+  sig { returns(Regexp) }
+  def formula_file_regex
+    @formula_file_regex ||= case formula_dir.basename.to_s
+    when "Formula"
+      %r{^Formula(/[^/]+)+\.rb$}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:760]: should this have two `+`?

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense. I do think the regex-based approach is a little harder to read so may e.g. warrant some interpolatated variables and comments.

File: Library/Homebrew/tap.rb (hunk 4 above)
Comment (apainintheneck) [reply Library/Homebrew/tap.rb:760]: Yep, it looks to be correct.

```rb
~ $ irb
irb(main):001:0> re = %r{^Formula(/[^/]+)+\.rb$}
=> /^Formula(\/[^\/]+)+\.rb$/
irb(main):002:0> re.match?("Formula/a/amber.rb")
=> true
irb(main):003:0> re.match?("Formula/amber.rb")
=> true
irb(main):004:0> re.match?("Formula.rb")
=> false
```

I can add some tests to make it clearer though.

**Comment by apainintheneck** [comment]
I could potentially try to use `File.fnmatch?` instead of regexes if that would be clearer.

File: Library/Homebrew/test/tap_spec.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/test/tap_spec.rb:204]: Ah, it doesn't work for two reasons.

1. `Tap#formula_file?` no longer supports `Pathname` objects.
2. `Tap#formula_file?` only supports relative paths from the Taps main directory.

**Review by carlocab — APPROVED** [review]

File: Library/Homebrew/tap.rb (hunk 2 above)
Comment (apainintheneck) [reply Library/Homebrew/tap.rb:773]: https://github.com/Homebrew/homebrew-test-bot/pull/1192

File: Library/Homebrew/tap.rb (hunk 4 above)
Comment (apainintheneck) [reply Library/Homebrew/tap.rb:760]: Tests were added in https://github.com/Homebrew/brew/pull/18167/commits/62a4734c9026500b77d30d092252b7e1ad59e6cc.

File: Library/Homebrew/tap.rb (hunk 3 above)
Comment (apainintheneck) [reply Library/Homebrew/tap.rb:764]: Addressed in https://github.com/Homebrew/brew/pull/18167/commits/62a4734c9026500b77d30d092252b7e1ad59e6cc

File: Library/Homebrew/tap.rb [hunk 5]
```diff
@@ -753,34 +753,34 @@ def cask_files_by_name
...
-    file.extname == ".rb"
+  sig { returns(Regexp) }
+  def formula_file_regex
+    @formula_file_regex ||= case formula_dir
+    when path/"Formula"
+      %r{^Formula(/[^/]+)+\.rb$}
+    when path/"HomebrewFormula"
+      %r{^HomebrewFormula(/[^/]+)+\.rb$}
+    when path
+      %r{^[^/]+\.rb$}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:764]:
```suggestion
    formula_file_regex = "[^/]+\.rb
    file_file_and_dir_regex = "(/[^/])*#{formula_file_regex}"
    when path/"Formula"
      %r{^Formula#{file_file_and_dir_regex}$}
    when path/"HomebrewFormula"
      %r{^HomebrewFormula#{file_file_and_dir_regex}$}
    when path
      %r{^#{formula_file_regex}$}
```
which possibly doesn't work at all but something like this that DRYs up the regex a bit and explains what's going on by breaking into components

Comment (apainintheneck) [reply Library/Homebrew/tap.rb:764]: I'm not a big fan of interpolating regexes based on smaller regexes especially when these regexes are already quite small because it ends up making things harder to read. Maybe adding a comment explaining each regex would be more helpful.

Comment (MikeMcQuaid) [reply Library/Homebrew/tap.rb:764]: I find the regexes as is pretty hard to follow. I don't think everything should always be DRY but, in this case, it is important that these three regexes have the same trailing end and only really differ by prefix. If you can find another way of achieving the same thing with better performance, cool, but, if not: I'd really like to split this into variables so that these don't run the risk of getting out of sync over time.

Comment (apainintheneck) [reply Library/Homebrew/tap.rb:764]: Tests have already been added to prevent this from happening so I don't think we need to worry about things getting out of sync here. IMO the biggest issue here is readability. Maybe something like this would be easier to read. It splits out the directory part of the regex into a separate section from the file name part and adds some inline comments. I haven't tested it yet but it should be close enough.

```rb
%r{
  ^Formula/
  (?:[^/]+/)* # zero or more subdirectories
  [^/]+\.rb   # ruby file name
}x
```

Comment (MikeMcQuaid) [reply Library/Homebrew/tap.rb:764]: > Tests have already been added to prevent this from happening

Tests don't prevent these getting out of sync, they just prevent them from getting out of sync in a way that breaks those specific tests.

> IMO the biggest issue here is readability.

I somewhat agree. I think the biggest issue is that you have three non-variable-assigned regexes and ⅔ of them differ only by prefix and they all have the same suffix. This harms both readability (it's hard to mentally parse where they are the same/not) but also: this is literally shared (and crucial!) logic that _must_ remain consistent (or things will go 💥). 

This is an almost textbook case of when you want to make things DRY. There's no world in which you want them to go out of sync.

> Maybe something like this would be easier to read. 

Do you propose to do these same comments for each of the regexes? This really seems like what variables are for...

> It splits out the directory part of the regex into a separate section from the file name part

Or: you could do this with variables? Please?

Comment (apainintheneck) [reply Library/Homebrew/tap.rb:764]: It sounds like you feel very strongly about this point so I'll update it. In the future, if you feel very strongly about something to the point that it's a blocker, could you use `Changed Requested` to better communicate that?

Comment (apainintheneck) [reply Library/Homebrew/tap.rb]: Addressed in https://github.com/Homebrew/brew/pull/18167/commits/60b8878f2176d7698bf789c5fdbbefc821b8ab05

**Comment by apainintheneck** [comment]
This is now ready for final review.

**Review by MikeMcQuaid — APPROVED** [review]
Thanks for making the regex changes! Looks good to merge now.

File: Library/Homebrew/tap.rb (hunk 5 above)
Comment (apainintheneck) [reply Library/Homebrew/tap.rb]: I would just like more clarity when it comes to which changes are nits/ignorable, suggestions/optional and blocking/required. It doesn't necessarily have to be expressed as `Changes Requested` but that's the most straightforward way to express that something is blocking/required.
